bip-musig2.mediawiki

```diff
@@ -315,15 +315,15 @@ Input:
 Input:
 * The secret signing key ''sk'': a 32-byte array or 0-byte array (optional argument)
 * The aggregate public key ''aggpk'': a 32-byte array or 0-byte array (optional argument)
-* The message ''m'': a 32-byte array or 0-byte array (optional argument)
+* The message ''m'': a byte array with ''0 &le; len(m) &le; 2<sup>32</sup>-1'' (optional argument)
```

Comment:
Now the arguments are somewhat inconsistent here. For `sk` and `aggpk`, the "0-byte array" corresponds to the argument not being present (a magic value in a sense). Whereas `m` can be present but the 0-byte array. For the concrete algorithm here, it doesn't really make a difference but it would be good to have a clear interface. 

Maybe just drop the "or 0-byte array" for `sk` and `aggpk` and handle the presence of the arguments explicitly in the pseudocode. This seems clearer anyway. A real implementation will probably also use an `Option`, or `NULL`/`nil` or whatever special thing the language offers here.

Reply:
Pushed a WIP commit. TODO: need to define what it means for an argument to be undefined (which looks quite terrible, we should rather "define" an optional type) and define the empty array.  Is this what you mean? Seems quite complex in our pseudocode language.

I didn't think that the interface was particularly unclear. If the message is present and 0 bytes, then provide it to the nonce function as such. If it's not present, provide a 0-byte array just like the other optional arguments.

Reply:
> Pushed a WIP commit. TODO: need to define what it means for an argument to be undefined (which looks quite terrible, we should rather "define" an optional type) and define the empty array. Is this what you mean? Seems quite complex in our pseudocode language.

Yep, this is what I meant, and I think it's okay as is. Suggestion: Just say in prose "If the optional argument `sk` is present" (or "defined" -- though I'd slightly prefer "present"). This makes the connection to the "optional argument" wording from the input list, and so it's certainly clear enough if you ask me. 

We could alternatively define an optional type but either your current version or my suggestion is good enough, so I think having an optional type is overkill for pseudocode (that's exactly a detail of concrete programming language that we're trying to abstract away from.) 

We should however use something idiomatic in the Python code, i.e., [`None` and `Optional`](https://stackoverflow.com/questions/51710037/how-should-i-use-the-optional-type-hint).   

> 
> I didn't think that the interface was particularly unclear. If the message is present and 0 bytes, then provide it to the nonce function as such. If it's not present, provide a 0-byte array just like the other optional arguments.



> I didn't think that the interface was particularly unclear. If the message is present and 0 bytes, then provide it to the nonce function as such. If it's not present, provide a 0-byte array just like the other optional arguments.

I think the algorithm is indeed clear but I could imagine myself writing a Python implementation and wondering whether I should accept `None` or reject it.

Reply:
Ok, I cleaned up that commit.

Reply:
Looks really clean now!